lib/rex/post/meterpreter/packet.rb

```diff
@@ -693,17 +819,31 @@ def to_r
   # passing it on to the default functionality that can parse
   # the TLV values.
   #
-  def from_r(bytes)
-    xor_key = bytes[0,4].unpack('N')[0]
-    super(xor_bytes(xor_key, bytes[4, bytes.length]))
+  def from_r(aes_key=nil)
+    xor_key = self.raw[XOR_KEY_OFF, XOR_KEY_SIZE]
+    data = xor_key + xor_bytes(xor_key, self.raw[SESSION_GUID_OFF, self.raw.length - SESSION_GUID_OFF])
+    self.session_guid = data[SESSION_GUID_OFF, SESSION_GUID_SIZE]
+    encrypted_flag = data[ENCRYPTED_FLAG_OFF, ENCRYPTED_FLAG_SIZE] == "\x01"
+    packet_length = data[PACKET_LENGTH_OFF, PACKET_LENGTH_SIZE]
+    packet_type = data[PACKET_TYPE_OFF, PACKET_TYPE_SIZE]
+    self.type = packet_type.unpack('N')[0]
+    if encrypted_flag
```

Comment:
Now that I'm thinking, doing away with trusting the `encrypted_flag` for sessions that we expect encrypted packets for is a hard problem.

@OJ said:
> After migration a session to another process, the negotiation phase needs to be re-run, because the AES key is not part of the configuration data (intentional). As a result, encrypt keys are cleared once migration is done and a new key is generated

This sounds like something a passive MitM can spoof at any stage and means we must not enforce the idea of "encrypted sessions must not play silly buggers with unencrypted packets". I shall noodle.